File: ios/Ezber/Models/StudySession.swift
import Foundation

/// Where the listener is in the verse × repetition queue. Saved at every repeat
/// boundary so the drill can always resume.
struct ResumePoint: Hashable, Codable {
    var verseNumber: Int
    var repeatIndex: Int

    init(verseNumber: Int, repeatIndex: Int) {
        self.verseNumber = verseNumber
        self.repeatIndex = max(1, repeatIndex)
    }
}

/// A run through a preset, with its resume point and completion counts.
struct StudySession: Identifiable, Hashable, Codable {
    var id: UUID
    var presetID: UUID
    var startedAt: Date
    var lastActiveAt: Date
    var resumePoint: ResumePoint
    var completedItems: Int
    var totalItems: Int
    var isCompleted: Bool

    init(
        id: UUID = UUID(),
        presetID: UUID,
        startedAt: Date = Date(),
        lastActiveAt: Date = Date(),
        resumePoint: ResumePoint,
        completedItems: Int = 0,
        totalItems: Int,
        isCompleted: Bool = false
    ) {
        self.id = id
        self.presetID = presetID
        self.startedAt = startedAt
        self.lastActiveAt = lastActiveAt
        self.resumePoint = resumePoint
        self.completedItems = completedItems
        self.totalItems = totalItems
        self.isCompleted = isCompleted
    }

    var completionRatio: Double {
        guard totalItems > 0 else { return 0 }
        return min(1, max(0, Double(completedItems) / Double(totalItems)))
    }
}

File: android/app/src/main/java/app/ezber/android/models/StudySession.kt
package app.ezber.android.models

/** Where the listener is in the verse × repetition queue. */
data class ResumePoint(val verseNumber: Int, val repeatIndex: Int) {
    init {
        require(repeatIndex >= 1) { "repeatIndex must be >= 1" }
    }
}

/**
 * A row in `sessions`: one run through a preset with its counters.
 * `plan_state` holds the exact resume position; this is the history record.
 */
data class StudySession(
    val id: Long = 0L,
    val presetId: Long,
    val startedAt: String,
    val endedAt: String? = null,
    val versesCovered: Int = 0,
    val repetitions: Int = 0,
    val interrupted: Boolean = false,
)

/**
 * The `plan_state` row for a preset: the exact resume position, written at
 * item (repeat) boundaries so a killed process resumes on the same repetition.
 */
data class PlanState(
    val presetId: Long,
    val planIndex: Int = 0,
    val positionMs: Long = 0L,
    val repetitionCountersJson: String? = null,
    val updatedAt: String? = null,
)
